File: feature_tests/java/somelib/src/main/java/dev/diplomattest/somelib/Utf16Wrap.java
package dev.diplomattest.somelib;

import dev.diplomattest.somelib.ntv.*;


import java.lang.foreign.Arena;
import java.lang.foreign.MemorySegment;
import java.lang.foreign.SegmentAllocator;
import java.lang.ref.Cleaner;
import java.util.List;
import static java.lang.foreign.ValueLayout.*;
import java.nio.charset.StandardCharsets;
import java.util.stream.Stream;

public class Utf16Wrap {

    MemorySegment internal;
    Cleaner.Cleanable cleanable;
    SegmentAllocator arena;

    List<Object> selfEdges = List.of();
    

    static class Utf16WrapCleaner implements Runnable {

        MemorySegment segment;
        Utf16WrapCleaner(MemorySegment segment) {
            this.segment = segment;
        }

        public void run() {
            somelib_h.Utf16Wrap_destroy(this.segment);
        }
    }

    Utf16Wrap() {}
    Utf16Wrap(MemorySegment handle, List<Object> selfEdges) {
        this.internal = handle;
        this.selfEdges = selfEdges;
        

    }
    
    public static Utf16Wrap fromUtf16(String input) {
        
        try (var arena = Arena.ofConfined()) {
            var inputData = Arena.ofAuto().allocateFrom(input, StandardCharsets.UTF_16);
            var inputLen = inputData.byteSize() - 1;  // allocated strings are null terminated
            var inputView = DiplomatString16View.allocate(Arena.ofAuto());
            DiplomatString16View.len(inputView, inputLen);
            DiplomatString16View.data(inputView, inputData);
            var nativeVal = somelib_h.Utf16Wrap_from_utf16(inputView);
            
            List<Object> selfEdges = List.of();
            
            
            
            var returnVal = new Utf16Wrap(nativeVal, selfEdges);
            return returnVal;
                    
        }
    }
    
    
    public String getDebugStr() {
        
        
        var writeable = somelib_h.diplomat_buffer_write_create(0);
        somelib_h.Utf16Wrap_get_debug_str(internal, writeable);
        var string = SliceUtils. readUtf8FromWriteable(writeable);
        somelib_h.diplomat_buffer_write_destroy(writeable);
        return string;
    }
    
    public String borrowCont() {
        
        try (var arena = Arena.ofConfined()) {
            
            
            var nativeVal = somelib_h.Utf16Wrap_borrow_cont(arena, internal);
            
            var returnVal = SliceUtils.readUtf16(nativeVal);
            return returnVal;
                    
        }
    }
    
}

File: feature_tests/kotlin/somelib/src/main/kotlin/dev/diplomattest/somelib/Utf16Wrap.kt
package dev.diplomattest.somelib;
import com.sun.jna.Callback
import com.sun.jna.Library
import com.sun.jna.Native
import com.sun.jna.Pointer
import com.sun.jna.Structure

internal interface Utf16WrapLib: Library {
    fun Utf16Wrap_destroy(handle: Pointer)
    fun Utf16Wrap_from_utf16(input: Slice): Pointer
    fun Utf16Wrap_get_debug_str(handle: Pointer, write: Pointer): Unit
    fun Utf16Wrap_borrow_cont(handle: Pointer): Slice
}

class Utf16Wrap internal constructor (
    internal val handle: Pointer,
    // These ensure that anything that is borrowed is kept alive and not cleaned
    // up by the garbage collector.
    internal val selfEdges: List<Any>,
)  {

    internal class Utf16WrapCleaner(val handle: Pointer, val lib: Utf16WrapLib) : Runnable {
        override fun run() {
            lib.Utf16Wrap_destroy(handle)
        }
    }

    companion object {
        internal val libClass: Class<Utf16WrapLib> = Utf16WrapLib::class.java
        internal val lib: Utf16WrapLib = Native.load("diplomat_feature_tests", libClass)
        @JvmStatic
        
        fun fromUtf16(input: String): Utf16Wrap {
            val inputSliceMemory = PrimitiveArrayTools.borrowUtf16(input)
            
            val returnVal = lib.Utf16Wrap_from_utf16(inputSliceMemory.slice);
            val selfEdges: List<Any> = listOf()
            val handle = returnVal 
            val returnOpaque = Utf16Wrap(handle, selfEdges)
            CLEANER.register(returnOpaque, Utf16Wrap.Utf16WrapCleaner(handle, Utf16Wrap.lib));
            inputSliceMemory?.close()
            return returnOpaque
        }
    }
    
    fun getDebugStr(): String {
        val write = DW.lib.diplomat_buffer_write_create(0)
        val returnVal = lib.Utf16Wrap_get_debug_str(handle, write);
        
        val returnString = DW.writeToString(write)
        return returnString
    }
    
    fun borrowCont(): String {
        
        val returnVal = lib.Utf16Wrap_borrow_cont(handle);
            return PrimitiveArrayTools.getUtf16(returnVal)
    }

}